lib/parser.ex

```diff
@@ -440,45 +440,157 @@ defmodule Antidote.Parser do
+      [clause] =
+        quote location: :keep do
+          unquote(int) -> escape_surrogate(unquote_splicing(args))
+        end
+      clause
+    end
+
+    defp escapeu_first_clause(int, first, last, rest, original, skip, stack, acc) do
+      skip = quote do: (unquote(skip) + 6)
+      acc =
+        quote bind_quoted: [acc: acc, first: first, last: last] do
+          [acc | <<((first <<< 8) + last)::utf8>>]
+        end
+      args = [rest, original, skip, stack, acc, 0]
+      [clause] =
+        quote location: :keep do
+          unquote(int) -> string(unquote_splicing(args))
+        end
+      clause
+    end
+
+    defmacro escapeu_last(int, original, skip) do
+      clauses = escapeu_last_clauses()
+      quote location: :keep do
+        case unquote(int) do
+          unquote(clauses ++ token_error_clause(original, skip, 6))
+        end
+      end
+    end
+
+    defp escapeu_last_clauses() do
+      for {int, last} <- unicode_escapes() do
+        [clause] =
+          quote do
+            unquote(int) -> unquote(last)
+          end
+        clause
+      end
+    end
+
+    defmacro escapeu_surrogate(int, last, rest, original, skip, stack, acc,
+             hi) do
+      clauses = escapeu_surrogate_clauses(last, rest, original, skip, stack, acc, hi)
+      quote location: :keep do
+        case unquote(int) do
+          unquote(clauses ++ token_error_clause(original, skip, 12))
+        end
+      end
+    end
+
+    defp escapeu_surrogate_clauses(last, rest, original, skip, stack, acc, hi) do
+      digits1 = 'Dd'
+      digits2 = Stream.concat([?C..?F, ?c..?f])
+      for {int, first} <- unicode_escapes(digits1, digits2) do
+        escapeu_surrogate_clause(int, first, last, rest, original, skip, stack,
+          acc, hi)
+      end
+    end
+
+    defp escapeu_surrogate_clause(int, first, last, rest, original, skip, stack,
+         acc, hi) do
+      skip = quote do: unquote(skip) + 12
+      acc =
+        quote bind_quoted: [acc: acc, first: first, last: last, hi: hi] do
+          lo = ((first &&& 0x03) <<< 8) + last
+          [acc | <<(hi + lo)::utf8>>]
+        end
+      args = [rest, original, skip, stack, acc, 0]
+      [clause] =
+        quote do
+          unquote(int) ->
+            string(unquote_splicing(args))
+        end
+      clause
     end
   end
-  defp escape_surrogate(<<_rest::bits>>, original, skip, _stack, _acc, _hi) do
-    error(original, skip + 6)
+
+  defp escapeu(<<int1::16, int2::16, rest::bits>>, original, skip, stack,
+       acc) do
+    require Unescape
+    last = escapeu_last(int2, original, skip)
+    Unescape.escapeu_first(int1, last, rest, original, skip, stack, acc)
+  end
+  defp escapeu(<<_rest::bits>>, original, skip, _stack, _acc) do
+    error(original, skip)
   end
 
-  defp try_codepoint(codepoint, string, position) do
-    <<codepoint::utf8>>
-  rescue
-    ArgumentError ->
-      token_error("\\u" <> string, position)
+  @compile {:inline, escapeu_last: 3}
```

Comment:
I will try the difference later. The second point seems doubtful because they are only used once so I'm unsure where the gain would come from, `escape_last` is only being inlined twice and has 22*22 clauses.